### DISCUSSION TITLE
How to convert SMILES to Amino Acids sequence using RDKit

### DISCUSSION THREAD

**Discussion by gundalav** [rdkit/rdkit#4659] [General] [answered]
I have the following SMILES, wrapped in Python RDKit:

    from rdkit import Chem
    m = Chem.MolFromSmiles('[H]N[C@@H](CCCCN)C(=O)N[C@@H](CCCCN)C(=O)N[C@@H](C)C(=O)N[C@@H](CC1=CNC2=C1C=CC=C2)C(O)=O'

Is there away I can convert that into an amino acid sequence, e.g, `KKAW`?

**Comment by dehaenw** [comment, marked as answer]
hello @gundalav.
in case there is MonomerInfo (so if you obtained the SMILES via Chem.MolFromSequence, MolFromFASTA etc) the problem is fairly simple. You can just use Chem.MolToSequence. However, if you do not have this info you will need to find the amino acids that are present, set the monomer info yourself and only then use MolToSequence. Below my prototype:
```python
import rdkit
from rdkit import Chem

def mol2seq(m):
    aa_smiles = {'ALA': 'C[C@H](N)C=O', 'CYS': 'N[C@H](C=O)CS', 'ASP': 'N[C@H](C=O)CC(=O)O', 'GLU': 'N[C@H](C=O)CCC(=O)O', 'PHE': 'N[C@H](C=O)Cc1ccccc1', 'GLY': 'NCC=O', 'HIS': 'N[C@H](C=O)Cc1c[nH]cn1', 'ILE': 'CC[C@H](C)[C@H](N)C=O', 'LYS': 'NCCCC[C@H](N)C=O', 'LEU': 'CC(C)C[C@H](N)C=O', 'MET': 'CSCC[C@H](N)C=O', 'ASN': 'NC(=O)C[C@H](N)C=O', 'PRO': 'O=C[C@@H]1CCCN1', 'GLN': 'NC(=O)CC[C@H](N)C=O', 'ARG': 'N=C(N)NCCC[C@H](N)C=O', 'SER': 'N[C@H](C=O)CO', 'THR': 'C[C@@H](O)[C@H](N)C=O', 'VAL': 'CC(C)[C@H](N)C=O', 'TRP': 'N[C@H](C=O)Cc1c[nH]c2ccccc12','TYR': 'N[C@H](C=O)Cc1ccc(O)cc1'}
    aas = ['GLY','ALA', 'VAL', 'CYS', 'ASP', 'GLU', 'PHE', 'HIS', 'ILE', 'LYS', 'LEU', 'MET', 'ASN', 'PRO', 'GLN', 'ARG', 'SER', 'THR', 'TRP','TYR'] #order important because gly is substructure of other aas
    # detect the atoms of the backbone and assign them with info
    CAatoms = m.GetSubstructMatches(Chem.MolFromSmarts("[C:0](=[O:1])[C:2][N:3]"))
    for atoms in CAatoms:
        a = m.GetAtomWithIdx(atoms[2])
        info = Chem.AtomPDBResidueInfo()
        info.SetName(" CA ") #spaces are important
        a.SetMonomerInfo(info)
    # detect the presence of residues and set residue name for CA atoms only
    for curr_aa in aas:
        matches = m.GetSubstructMatches(Chem.MolFromSmiles(aa_smiles[curr_aa]))
        for atoms in matches:
            for atom in atoms:
                a = m.GetAtomWithIdx(atom)
                info = Chem.AtomPDBResidueInfo()
                if a.GetMonomerInfo() != None:
                    if a.GetMonomerInfo().GetName() == " CA ":
                        info.SetName(" CA ")
                        info.SetResidueName(curr_aa)
                        a.SetMonomerInfo(info)
    # renumber the backbone atoms so the sequence order is correct:
    bbsmiles = "O"+"C(=O)CN"*len(m.GetSubstructMatches(Chem.MolFromSmiles(aa_smiles["GLY"]))) # generate backbone SMILES
    backbone = m.GetSubstructMatches(Chem.MolFromSmiles(bbsmiles))[0]
    id_list = list(backbone)
    id_list.reverse()
    for idx in [a.GetIdx() for a in m.GetAtoms()]:
        if idx not in id_list:
            id_list.append(idx)
    m_renum = Chem.RenumberAtoms(m,newOrder=id_list)
    return Chem.MolToSequence(m_renum)


mol=Chem.MolFromSmiles("CC[C@H](C)[C@H](NC(=O)[C@H](CCCCN)NC(=O)[C@H](CC(=O)O)NC(=O)[C@@H](N)CCCNC(=N)N)C(=O)N[C@H](C(=O)O)[C@@H](C)O") #seq Without monomerinfo
print(mol2seq(mol))
```

which returns the correct one letter AA sequence, in this case
```
RDKIT
```

Apologies in advance for the somewhat messy code, I did not have time to clean it up or optimize it. However, it seems to work well enough for sequences with canonical AAs.

**Reply by gundalav** [reply]
@dehaenw Thanks a million! Works like charm!
BTW, can your code be used for [multiple dialects of SMILES](https://open-babel.readthedocs.io/en/latest/FileFormats/SMILES_format.html?highlight=smiles) (universal, inchified, canonical, etc), which may represent the same structure with a different string?

Or is it only for universal form? So if with other form (e.g. inchified, canonical) I may have to convert them first.

**Reply by dehaenw** [reply]
It should be independent of the atom numbering of the mol because the backbone atom ids get renumbered anyway. But some tests would need to be run if that's actually the case, I haven't checked it thoroughly myself. Also, there was a mistake with the renumbering part of the script, i fixed that now.

**Comment by merritjm** [comment]
Hi, I found your mol2seq code very useful.  I'm interested in non-natural amino acids.  Aib for example.  When i pass the smiles string for Aib through your code it incorrectly represents it as 'A' or Alanine.  The substructures are very close.  I added a key for it in the aa_smiles and aas lists, and then the code outputs the identifier as 'X'.  I'm guessing this is a default in rdkit because the substructure is not recognized.  Is there a way to supply this information somehow?
Thanks,

**Reply by dehaenw** [reply]
hi @merritjm,
i think even if there are no explicit definitions of various non natural amino acids it should probably be possible to identify them by substructure search and replace the X with whatever symbol you pick for it. The function I posted here earlier is also imperfect because as you detected it can give false matches when there is a substructure match that doesn't cover the entire residue or even when the aa definitions are shuffled. when i have some time in a few days i'll post an updated version here or any problems i run into trying to do it!

**Comment by mirfanmalik** [comment]
How to convert 9 Amino Acids sequence to  SMILES using RDKit
Can we generate the descriptors using 9 aminoacid sequence
e.g. If I want to pass 'VEVFAQNMV'
Is it possible that instead of smile if I pass only smile sequence and RDkit generate the descriptors against it?
if not! is there any solution to convert the given sequence (VEVFAQNMV) to a smile using python function ???
Please I need your help in this regard
thanks

**Reply by dehaenw** [reply]
```python
from rdkit import Chem

Chem.MolToSmiles(Chem.MolFromFASTA("VEVFAQNMV"))
```

outputs the string `CSCC[C@H](NC(=O)[C@H](CC(N)=O)NC(=O)[C@H](CCC(N)=O)NC(=O)[C@H](C)NC(=O)[C@H](Cc1ccccc1)NC(=O)[C@@H](NC(=O)[C@H](CCC(=O)O)NC(=O)[C@@H](N)C(C)C)C(C)C)C(=O)N[C@H](C(=O)O)C(C)C`

be careful calculating certain descriptors (e.g. Morgan counts) because they are often intended for and tested on small molecules and not peptides so can give meaningless results for oligopeptides.

**Reply by mirfanmalik** [reply]
thanks @dehaenw
